assignment-client/src/assets/AssetServer.cpp

```diff
@@ -0,0 +1,279 @@
+//
+//  AssetServer.cpp
+//
+//  Created by Ryan Huffman on 2015/07/21
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+
+#include "AssetServer.h"
+
+#include <QBuffer>
+#include <QCryptographicHash>
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QCoreApplication>
+#include <QEventLoop>
+#include <QRunnable>
+#include <QString>
+
+#include <NodeType.h>
+
+const QString ASSET_SERVER_LOGGING_TARGET_NAME = "asset-server";
+
+void writeError(NLPacketList* packetList, AssetServerError error) {
+    packetList->writePrimitive(error);
+}
+
+class SendAssetTask : public QRunnable {
+public:
+    SendAssetTask(MessageID messageID, const QByteArray& assetHash, QString filePath, DataOffset start, DataOffset end, const SharedNodePointer& sendToNode) :
+        QRunnable(),
+        _messageID(messageID),
+        _assetHash(assetHash),
+        _filePath(filePath),
+        _start(start),
+        _end(end),
+        _sendToNode(sendToNode)
+    {
+    }
+
+    void run();
+
+signals:
+    void finished();
+
+private:
+    MessageID _messageID;
+    QByteArray _assetHash;
+    QString _filePath;
+    DataOffset _start;
+    DataOffset _end;
+    SharedNodePointer _sendToNode;
+};
+
+void SendAssetTask::run() {
+    qDebug() << "Starting task to send asset: " << _assetHash << " for messageID " << _messageID;
+    auto replyPacketList = std::unique_ptr<NLPacketList>(new NLPacketList(PacketType::AssetGetReply, QByteArray(), true, true));
+
+    replyPacketList->write(_assetHash, HASH_HEX_LENGTH);
+
+    replyPacketList->writePrimitive(_messageID);
+
+    const int64_t MAX_LENGTH = 4294967296;
+
+    if (_end <= _start) {
+        writeError(replyPacketList.get(), AssetServerError::INVALID_BYTE_RANGE);
+    } else if (_end - _start > MAX_LENGTH) {
+        writeError(replyPacketList.get(), AssetServerError::INVALID_BYTE_RANGE);
+    } else {
+        QFile file { _filePath };
+        qDebug() << "Opening file: " << QString(QFileInfo(_assetHash).fileName());
+
+        if (file.open(QIODevice::ReadOnly)) {
+            if (file.size() < _end) {
+                writeError(replyPacketList.get(), AssetServerError::INVALID_BYTE_RANGE);
+            } else {
+                auto size = _end - _start;
+                file.seek(_start);
+                replyPacketList->writePrimitive(AssetServerError::NO_ERROR);
+                replyPacketList->writePrimitive(size);
+                while (file.pos() < file.size()) {
+                    static const int chunkSize = 20000;
+                    QByteArray data = file.read(chunkSize);
+                    replyPacketList->write(data, chunkSize);
```

Comment:
The read should work, the size passed is an upper limit. That said, the chunkSize will be incorrect on write, although it's not needed like you mentioned.

In any case, this was a temporary hack because of the recursive nature of PacketList::writeData - that is fixed now, so I will remove this.